src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcWebSocketServerTransport.cs

```diff
@@ -0,0 +1,146 @@
+        ParseWebSocketURL(address, out _endPoint);
+        _cancellation = new CancellationTokenSource();
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            _cancellation.Cancel();
+
+            singleton.DropRef();
+
+            _cancellation.Dispose();
+        }
+    }
+
+    public override async Task<Stream> AcceptAsync(CancellationToken token)
+    {
+        if (singleton.AddRef())
+        {
+            await singleton.StartServer(_endPoint, token);
+        }
+        Stream s = await singleton.AcceptConnection(token);
+        return s;
+    }
+
+    private static void ParseWebSocketURL(string endPoint, out Uri uri)
+    {
+        string uriToParse;
+        // Host can contain wildcard (*) that is a reserved charachter in URI's.
+        // Replace with dummy localhost representation just for parsing purpose.
+        if (endPoint.IndexOf("//*", StringComparison.Ordinal) != -1)
+        {
+            // FIXME: This is a workaround for the fact that Uri.Host is not set for wildcard host.
+            throw new ArgumentException("Wildcard host is not supported for WebSocket endpoints");
+        }
+        else
+        {
+            uriToParse = endPoint;
+        }
+
+        string[] supportedSchemes = new string[] { "ws", "wss", "http", "https" };
+
+        if (!string.IsNullOrEmpty(uriToParse) && Uri.TryCreate(uriToParse, UriKind.Absolute, out uri))
+        {
+            bool supported = false;
+            foreach (string scheme in supportedSchemes)
+            {
+                if (string.Compare(uri.Scheme, scheme, StringComparison.InvariantCultureIgnoreCase) == 0)
+                {
+                    supported = true;
+                    break;
+                }
+            }
+            if (!supported)
+            {
+                throw new ArgumentException(string.Format("Unsupported Uri schema, \"{0}\"", uri.Scheme));
+            }
+            return;
+        }
+        else
+        {
+            throw new ArgumentException(string.Format("Could not parse {0} into host, port", endPoint));
+        }
+    }
+
+
+    // a coordination class that ensures we only start a single webserver even as the
+    // diagnostic server protocol requires us to accept multiple connections.
+    // while it is alive, each connection owns one reference to this singleton.
+    // when the reference count goes from 0->1, we start the server.
+    // when the references drops back to zero, we let the webserver stop.
+    internal class Singleton
+    {
+        private volatile int _refCount;
+        public bool ServerRunning { get; internal set; } = false;
+        public bool ServerStopping { get; internal set; } = false;
+        public WebSocketServer.IWebSocketServer server { get; internal set; } = null;
+        internal Singleton()
+        {
+            _refCount = 0;
+        }
+
+        public bool AddRef()
```

Comment:
yea there are concurrency problems here.

I've been thinking more about how this should work for a scenario where we might use this in a library (as part of the WasmAppHost, for example for `dotnet run`) - it might be better to control the lifetime of the server from the main dotnet-dsrouter process and to just give IpcWebSocketServerTransport a handle to it.  That's more in line with how the library scenario would work - the `dotnet run` would be starting the server and adding the routing, and the DiagnosticsServerRouter would just handle the connections.  

I will update this part

Reply:
Ok, this is done now.  The dotnet-dsrouter program now controls the lifetime of the embedded web server.  The diagnostic client library now just asks for an instance